dvc/cache/base.py

```diff
@@ -436,6 +444,10 @@ def _checkout_file(
         if progress_callback:
             progress_callback(str(path_info))
 
+        if isexec:
+            st = os.stat(path_info.fspath)
+            os.chmod(path_info.fspath, st.st_mode | stat.S_IEXEC)
+
```

Comment:
IIRC chmod won't modify the mtime(we use it to detect changes in the workspace), so there is nothing cache should worry about when doing chmod +x. The chmod part is more related to the output itself, than to the cache (notice how we are just passing through the isexec, but don't really use anything cache-specific). That's why the saving of isexec is not in `cache.save` but is actually in `output.save`. So let's move this to `OutputBase.checkout`, right after `cache.checkout` call. Also note that it is better to use tree.* methods instead of `os.*`, as that helps to keep the abstraction for different types of outputs :slightly_smiling_face:

Reply:
Good point! I moved this code to `BaseOutput.checkout` and used `tree.*` methods. There is one minor stylistic issue: all methods in `BaseTree` are using `path_info` as an argument, I've added `BaseTree.isexec(path)` since it was not there and all children classes use `path` argument in that method:

https://github.com/iterative/dvc/blob/2ade7f7ba56c1b76924f9b5b7f379225b89e1914/dvc/tree/base.py#L198

Reply:
I modified isexec to use `path_info` as an argument